### PR TITLE
avoid promoting division, modulo and indexing operations that could fail

### LINKED ISSUE #61821: Duplicated errors on UB inside a promoted
The following output should only emit the error once:
```rust
#![warn(const_err)]

fn main() {
    &{ [1, 2, 3][4] };
}
```

```
error: index out of bounds: the len is 3 but the index is 4
  --> $DIR/array-literal-index-oob.rs:2:7
   |
LL |     &{[1, 2, 3][4]};
   |       ^^^^^^^^^^^^
   |
   = note: `#[deny(const_err)]` on by default

error: reaching this expression at runtime will panic or abort
  --> $DIR/array-literal-index-oob.rs:2:7
   |
LL |     &{[1, 2, 3][4]};
   |     --^^^^^^^^^^^^-
   |       |
   |       index out of bounds: the len is 3 but the index is 4

error: aborting due to 2 previous errors
```

_Follow up to https://github.com/rust-lang/rust/pull/61598_

### PR DESCRIPTION
**Pull request by RalfJung**
For division, `x / y` will still be promoted if `y` is a non-zero integer literal; however, `1/(1+1)` will not be promoted any more.

While at it, also see if we can reject promoting floating-point arithmetic (which are [complicated](https://github.com/rust-lang/unsafe-code-guidelines/issues/237) so maybe we should not promote them).

This will need a crater run to see if there's code out there that relies on these things being promoted.

If we can land this, promoteds in `fn`/`const fn` cannot fail to evaluate any more, which should let us do some simplifications in codegen/Miri!

Cc https://github.com/rust-lang/rfcs/pull/3027
Fixes https://github.com/rust-lang/rust/issues/61821
r? @oli-obk

### CODE REVIEW

File: src/test/ui/consts/const-eval/promoted_errors.rs [hunk 1]
```diff
@@ -8,21 +8,34 @@
...
+}
+const fn oob() -> i32 {
+    [1,2,3][4]
+}
+
+const X: () = {
+    let _x: &'static u32 = &overflow();
+    //[opt_with_overflow_checks,noopt]~^ WARN any use of this value will cause an error
+    let _x: &'static i32 = &div_by_zero1();
+    //[opt]~^ WARN any use of this value will cause an error
```
Comment (RalfJung) [inline src/test/ui/consts/const-eval/promoted_errors.rs:34]: Curiously, the errors stop after the first panic. This should change if we decide that promoteds get added to (or treated like) `required_consts`.

File: compiler/rustc_mir/src/transform/promote_consts.rs [hunk 2]
```diff
@@ -626,9 +649,28 @@ impl<'tcx> Validator<'_, 'tcx> {
...
+                        if !self.explicit {
+                            // The RHS must be a non-zero const.
+                            let const_val = match rhs {
+                                Operand::Constant(c) => {
+                                    c.literal.try_eval_bits(self.tcx, self.param_env, lhs_ty)
+                                }
+                                _ => None
+                            };
+                            match const_val {
+                                Some(x) if x != 0 => {} // okay
```
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/promote_consts.rs:668]: This is very pedantic and not related to PR logic at all, but isn't this a great situation for

```suggestion
                                Some(1..) => {} // okay
```

File: src/test/ui/consts/const-eval/promoted_errors.rs (hunk 1 above)
Comment (oli-obk) [reply src/test/ui/consts/const-eval/promoted_errors.rs:34]: hmm... interesting. We must have an early abort somewhere in const prop

File: compiler/rustc_mir/src/transform/promote_consts.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_mir/src/transform/promote_consts.rs:668]: This is `x > 0`, not `x != 0`, which is only equivalent for unsigned types, and it is not locally obvious that `x` is unsigned. So I think I'd prefer to stick to `!= 0`.

File: src/test/ui/consts/const-eval/promoted_errors.rs (hunk 1 above)
Comment (RalfJung) [reply src/test/ui/consts/const-eval/promoted_errors.rs:34]: No, ConstProp is not involved I think. The error is emitted by ConstEval. And of course ConstEval stops on the first error.

**Comment by RalfJung** [comment]
@craterbot check

**Comment by RalfJung** [comment]
Non-spurious regressions:
* float operations:
   https://crater-reports.s3.amazonaws.com/pr-80579/try%23575650af7d7ca88602df28e39d27bf3e456c306e/gh/WimbledonLabs.graphics_vid/log.txt
    https://crater-reports.s3.amazonaws.com/pr-80579/try%23575650af7d7ca88602df28e39d27bf3e456c306e/reg/nyx-space-0.0.23/log.txt
    https://crater-reports.s3.amazonaws.com/pr-80579/try%23575650af7d7ca88602df28e39d27bf3e456c306e/gh/finegeometer.four-dimensions-2/log.txt
    https://crater-reports.s3.amazonaws.com/pr-80579/try%23575650af7d7ca88602df28e39d27bf3e456c306e/gh/tomc1998.rs-proc-gen-experiment/log.txt
    https://crater-reports.s3.amazonaws.com/pr-80579/try%23575650af7d7ca88602df28e39d27bf3e456c306e/reg/fast-math-0.1.1/log.txt
  So, I'll back out the float part of this. Float division AFAIK never fails, so I'll keep the "not `0`" test only for ints.
* array indexing:
 https://crater-reports.s3.amazonaws.com/pr-80579/try%23575650af7d7ca88602df28e39d27bf3e456c306e/reg/multiboot-0.6.0/log.txt
  Since array lengths are statically known, I can permit indexing into them with a literal.

**Comment by RalfJung** [comment]
Okay, I think this adjustment should take care of all the regressions.

File: compiler/rustc_mir/src/transform/promote_consts.rs [hunk 3]
```diff
@@ -628,7 +687,22 @@ impl<'tcx> Validator<'_, 'tcx> {
                 }
 
                 match op {
-                    // FIXME: reject operations that can fail -- namely, division and modulo.
+                    BinOp::Div | BinOp::Rem => {
+                        if !self.explicit && lhs_ty.is_integral() {
+                            // Integer division: the RHS must be a non-zero const.
+                            let const_val = match rhs {
+                                Operand::Constant(c) => {
+                                    c.literal.try_eval_bits(self.tcx, self.param_env, lhs_ty)
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/promote_consts.rs:695]: The one concern I have with this is that it could make promotion dependent on the value of a constant in another crate. I wonder if it is worth to somehow ensure that this is either a literal or a crate-local constant.

Comment (oli-obk) [reply compiler/rustc_mir/src/transform/promote_consts.rs:691]: I think it's fine. If the constant is evauable at this point, then it must be part of the public API of the other crate. Changing it can already cause breaking changes without taking promotion into account. The diagnostic is just a bit more suprising with promotion

**Comment by oli-obk** [comment]
Since this is a breaking change (although not breaking anything in practice, and the code had to already contain broken things or weird usages of `cfg` in order to break now), this needs sign off by @rust-lang/lang 

TLDR: we now do not promote fallible operations (unless their use is guaranteed to not fail). This causes `&(1/0)` to not get promoted anymore and thus not have the `'static` lifetime. More notably this means that `&(1 / T::N)` is not promoted anymore, because we don't know whether `T::N` is `0`. Any use of such a promoted would have failed anyway, but they could have existed in dead code. So there could be some dead code in private repos that breaks due to this change.

**Comment by RalfJung** [comment]
> More notably this means that &(1 / T::N) is not promoted anymore, because we don't know whether T::N is 0. Any use of such a promoted would have failed anyway, but they could have existed in dead code. So there could be some dead code in private repos that breaks due to this change.

Well, if the code was only ever monomorphized with `T::N` not being 0, it actually could have been live code as well that will break now.

File: compiler/rustc_mir/src/transform/promote_consts.rs [hunk 4]
```diff
@@ -498,16 +496,55 @@ impl<'tcx> Validator<'_, 'tcx> {
...
+                                None
+                            };
+                            if let Some(idx) = val {
+                                // Determine the type of the thing we are indexing.
+                                let ty = place_base.ty(self.body, self.tcx).ty;
+                                match ty.kind() {
+                                    ty::Array(_, len) => {
+                                        // It's an array; determine its length.
+                                        if let Some(len) =
+                                            len.try_eval_usize(self.tcx, self.param_env)
```
Comment (Aaron1011) [inline compiler/rustc_mir/src/transform/promote_consts.rs:528]: Will this work even if `len` is a constant defined in another crate? If so, this could lead to extremely confusing errors in the unlikely event that the upstream crate changes the value to something that makes us unable to perform promotion.

Comment (RalfJung) [reply compiler/rustc_mir/src/transform/promote_consts.rs:528]: Yes, it will, and I wondered about the same thing: https://github.com/rust-lang/rust/pull/80579#discussion_r553263361.

**Comment by nikomatsakis** [comment]
We discussed this in the [2021-01-12](https://github.com/rust-lang/lang-team/blob/master/minutes/2021-01-12.md) meeting and the consensus was that we should land it. Approved by @rust-lang/lang!

**Comment by oli-obk** [comment]
r=me after a rebase

File: src/test/ui/consts/const-eval/promoted_errors.rs [hunk 5]
```diff
@@ -8,21 +8,34 @@
 
 #![warn(const_err, arithmetic_overflow, unconditional_panic)]
 
+// The only way to have promoteds that fail is in `const fn` called from `const`/`static`.
+const fn overflow() -> u32 {
+    0 - 1
```
Comment (RalfJung) [inline src/test/ui/consts/const-eval/promoted_errors.rs:13]: @oli-obk unlike in previous versions of this PR, the "arithmetic overflow" and "unconditional panic" lint does not fire here any more. Any idea why that is?

Comment (RalfJung) [reply src/test/ui/consts/const-eval/promoted_errors.rs:13]: Ah, I guess this is because we now have separate CTFE MIR, and do not run const-prop any more on the CTFE MIR. The `fn` is private and otherwise unused so regular MIR is never created.
